### PR TITLE
Adds a new "UI and gem signin" mfa_level to allow API Keys to skip MFA check

### LINKED ISSUE #2500: MFA and API Keys in CI/CD
I'm having issues with Rubygems MFA.
I've previously generated an API Key that I've been using in my CI/CD pipelines to publish new versions of gems.

However, after recently enabling 2FA/MFA, the pipeline has started failing in Travis with the following error:
`Access Denied. Please sign up for an account at https://rubygems.org (Gems::GemError)`

Although I completely agree with 2FA and MFA, I feel like the current implementation doesn't take into consideration automation and CI/CD Pipelines.
In my experience with other services, 2FA and MFA usually do not apply to API Keys. i.e. when using an API Key to authenticate (which is supposed to be long and hard to hack), you can bypass the 2FA/MFA check.

The current options, from what I read in the doc, are the following, but both of them are not good in the case of CI/CD Pipelines:

1. Disable 2FA/MFA checks for API and leave it on for UI only: this is of course insecure and nullifies the extra security as it seems like the CLI uses the API and you can authenticate to the CLI using less secure means other than API Keys (see below for more information).
2. Use the command line `--totp` argument: this works with CLI, but unfortunately is not possible to the top of my knowledge in CI/CD Pipelines, as I have no idea on how to make them sync with the Authenticator App on my phone.

The confusion on point 1 comes with the CLI.
On my laptop, the first time I used the cli I remember my username and password were asked:

> After creating the account, use your email and password when pushing the gem. (RubyGems saves the credentials in ~/.gem/credentials for you so you only need to log in once.)

Now, that is of course unsafe, so I'd ideally like the CLI to request the MFA code in order to do anything. However, according to the documentation, that is only possible by enabling MFA for both UI and API (https://guides.rubygems.org/using-mfa-in-command-line/).

My suggestion would be to provide an extra option so that 2FA/MFA can be enabled for BOTH API and UI, but ignored when API authentication is done through API Keys only.

Any help would be greatly appreciated!

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [x] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

### PR DESCRIPTION
**Pull request by iMacTia**
Adds a new "UI and gem signin" mfa_level to allow API Keys to skip MFA check
Fixes #2500

# Additional Notes
Apologies for the auto-formatting, my IDE decided indentation was better this way but since Rubocop didn't complain I left it that way, I hope that's fine

### CODE REVIEW

**Comment by sonalkr132** [comment]
You need to update the api endpoints used in gem signin to use this new level [[1](https://github.com/rubygems/rubygems.org/blob/master/app/controllers/api/v1/api_keys_controller.rb#L50)]. 

Also, please don't make spacing changes.

**Comment by iMacTia** [comment]
Thanks @sonalkr132, it wasn't that easy to find without some help!
Added a commit, hopefully this is everything needed now

**Review by sonalkr132 — APPROVED** [review]
This looks good :+1: 
A couple of things..
* if possible break the test function you have added into smaller functions. It is fine as it is.. so maybe just add the file name to rubycop exclusion. we need to make the build green before we can merge https://travis-ci.org/github/rubygems/rubygems.org/jobs/753702130#L919
* update the guide for this to mention the new level and any clarification you may have regarding comparison with "UI only" level https://github.com/rubygems/guides/blob/gh-pages/setting-up-multifactor-authentication.md

**Comment by iMacTia** [comment]
Thank you @sonalkr132 and apologies for the delay!
Gave it the final push and opened https://github.com/rubygems/guides/pull/277 as well for the documentation 🙌

File: test/functional/api/v1/api_keys_controller_test.rb [hunk 1]
```diff
@@ -50,6 +50,116 @@ def self.should_deny_access_missing_otp
     end
   end
 
+  def should_return_api_key_successfully
```
Comment (sonalkr132) [inline test/functional/api/v1/api_keys_controller_test.rb:53]: > undefined method `should_return_api_key_successfully' for Api::V1::ApiKeysControllerTest:Class (NoMethodError)

I think self is missing.

**Comment by sonalkr132** [comment]
Thank you for working on this :star2: This should be deployed in a couple of days.

**Comment by iMacTia** [comment]
@sonalkr132 Thanks to you for your patience and support ❤️ !

**Comment by iMacTia** [comment]
@sonalkr132 finally managed to try this today and the push via GitHub Action worked successfully with 2FA and "UI and gem signing" option 🎉

**Comment by ioquatix** [comment]
This is a really great first step.

I think ideally what I'd like is:

- API key specific to one or more gems.
- Enable/disable OTP requirement per key.
- Ability to specify key via command line e.g. ENV.

**Comment by sonalkr132** [comment]
Good ideas. Please open an issue with it.

> Ability to specify key via command line e.g. ENV.

if you mean otp key, this is already supported. you can use --otp or GEM_HOST_OTP_CODE https://github.com/rubygems/rubygems/pull/4697. API key also has --api-key or GEM_HOST_API_KEY

**Pull request merged**